File: cpp/046.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> out;
        help(nums, 0, out);
        return out;
    }
    
    void help(vector<int>& nums, int start, vector<vector<int>> &out) {
        if (start == nums.size()) {
            out.push_back(nums);
            return;
        }
        int size = nums.size();
        for (int i = start; i < size; ++i) {
            swap(nums[i], nums[start]);
            help(nums, start + 1, out);
            swap(nums[i], nums[start]);
        }
    }
};

File: go/046.go
func permute(nums []int) [][]int {
    var out [][]int
    help(nums, 0, &out)
    return out
}

func help(nums []int, start int, out *[][]int) {
    if start == len(nums) {
        temp := make([]int, len(nums))
        copy(temp, nums)
        *out = append(*out, temp)
        return
    }
    size := len(nums)
    for i := start; i < size; i++ {
        nums[i], nums[start] = nums[start], nums[i]
        help(nums, start+1, out)
        nums[i], nums[start] = nums[start], nums[i]
    }
}

